### ISSUE TITLE
Reduce allocations in `XamlMember.ToString`

### ISSUE THREAD

**Issue by Youssef1313** [unoplatform/uno#13232]
![image](https://github.com/unoplatform/uno/assets/31348972/61739a09-96ce-44aa-b1a1-3dd037345430)

![image](https://github.com/unoplatform/uno/assets/31348972/1ccc8801-53be-47db-b065-1f21d04e77fb)

**Comment by KevinFerati** [comment]
For reference :

![image](https://github.com/unoplatform/uno/assets/71434466/e2d172bb-ea80-401b-9960-649abc9da1b5)

Before diving in a PR, I'll put my analysis here.

The issue arises with `String.Concat(params string[])` in the else branch, which allocates two string[] (one when passing the arguments, the other one in the function itself). I've played around with some solutions  : using a StringBuilder with a predefined capacity, using a stack and heap-allocated char buffer (which are then filled with the strings by copying them character by character).

Here's a (quick and ugly) benchmark comparing them, each returning the `XamlMember.ToString`  string in somewhat the same format with small strings: 

![image](https://github.com/unoplatform/uno/assets/71434466/4f6c5cd5-20af-4acc-86de-8e868fb0ce51)

Here's one with bigger strings for good measure : 
![image](https://github.com/unoplatform/uno/assets/71434466/33a99994-6aef-421d-9cbb-71950942920d)

The stack-allocated version has ~1/3 to 15% less allocations, but takes more time, probably due to the manual copying from the source strings. Would that be an issue ? Anyway, I'll try playing around with `System.Buffer.MemoryCopy()` or something else to see whether this helps or not. 

I can gladly give the source for my benchmarks and the strings i've used if anyone needs it :)

**Comment by agneszitte** [comment]
> Before diving in a PR, I'll put my analysis here.

fyi @jeromelaban, @Youssef1313

**Comment by KevinFerati** [comment]
> > Before diving in a PR, I'll put my analysis here.
> 
> thanks a lot for the details @KevinFerati!
> 
> fyi @jeromelaban, @Youssef1313, @ebariche
De rien ! 

Also, to add to my initial post : I've tried benchmarking the difference in the number of allocated bytes between a litteral string and one created by a stack-allocated char buffer or by mutating directly a newly created string. Turns out out that they have the same number of allocated bytes, so I don't think we can go below that threshold.

**Comment by jeromelaban** [comment]
Those are interesting findings, thanks.

Looking at the code further `XamlMember.ToString()` should not necessarily be used at all in [BuildExtendedProperties](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlFileGenerator.cs#L3003-L3004), as it is not supposed represent a value that can be used in the generated code (other than for diagnostics). I wonder if we should be removing its use altogether. Finding its actual uses would be better than fixing the allocation problem.

**Comment by KevinFerati** [comment]
`XamlMember.ToString()` is used by its `GetHashCode()`  which, according to #13228 , is in turn called by the underlying Dictionary when it's memoized.  Another Dictionary with a key of type `XamlMember` can be found in [ParsedMarkupExtensionInfo](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/System.Xaml/System.Xaml/ParsedMarkupExtensionInfo.cs)  (though it seems like a lot of its keys could be XamlDirectives, but i'm not quite sure).

Anyway, basically everything that uses the hash code of a `XamlMember`.

**Comment by jeromelaban** [comment]
Right. There are multiple ways to think about this:
1. We could cache [ToString()](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/System.Xaml/System.Xaml/XamlMember.cs#L295), though it would still be generating strings
1. We could modify [GetHashCode](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/System.Xaml/System.Xaml/XamlMember.cs#L289) to create a hash based on the presence of a [DeclaringType](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/System.Xaml/System.Xaml/XamlMember.cs#L299) without creating strings, then cache the hashcode value as `XamlMember` seems to effectively be immutable.
1. Change the [XamlMember redirection type](https://github.com/unoplatform/uno/blob/3b918ce59ff8fd84cf0e5a59e7bbf0050c36b536/src/SourceGenerators/Uno.UI.SourceGenerators/XamlGenerator/XamlRedirection/XamlMember.cs#L69) to do something similar as point 2.

I'd likely prefer solution 2.

**Comment by MartinZikmund** [comment]
Relevant
